### PR TITLE
Orders > Products : Display content of a pack

### LINKED ISSUE #16393: Missing button to display the content of a pack in the new order page
**Describe the bug**
The button to display the content of a pack in the new order page is missing

**To Reproduce**
1. Go to FO and order the product "Pack Mug + affiche encadrée"
2. Display the order in BO
3. There isn't the button "Package content" to display the content of the pack

**Screenshots**
capture d'écran_2147

Legacy

capture d'écran_2146

**New design**

A link ("View pack content") is displayed below the products in a pack on product list:
Capture d’écran 2019-11-28 à 15 40 06

It opens the products from the pack in a popup: 
Capture d’écran 2019-11-28 à 15 43 29

**Additional information**
PrestaShop version: develop
PHP version: N/A

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | "Orders > Products" : Display content of a pack
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #16393
| How to test?  | Check beahavior in BO > Orders > Product when you have a pack (design in  #16393)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16722)

### CODE REVIEW

File: admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js [hunk 1]
```diff
@@ -105,6 +105,17 @@ export default {
   productEditLocationText: '.editProductLocation',
   productEditAvailableText: '.editProductAvailable',
   productEditTotalPriceText: '.editProductTotalPrice',
+  // Product Pack Modal
+  productPackModal: '#product-pack-modal',
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js]: You should be able to create another level, really easier to read :sweat_smile: 
```suggestion
  productPackModal: {
    modla: '#product-pack-modal',
    table: '#product-pack-modal-table tbody',
    rows: '#product-pack-modal-table tbody tr:not(#template-pack-table-row)',
    template: '#template-pack-table-row',
    product: {
      img: '.cell-product-img img',
      name: '.cell-product-name .product-name',
      ref: '.cell-product-name .product-reference',
      supplierRef: '.cell-product-name .product-supplier-reference',
      quantity: '.cell-product-quantity',
      availableQuantity: '.cell-product-available-quantity',
    },
  },
```

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php [hunk 2]
```diff
@@ -438,8 +438,42 @@ private function getOrderProducts(Order $order): OrderProductsForViewing
...
             $orderInvoice = new OrderInvoice($product['id_order_invoice']);
 
+            $packItems = [];
+            foreach ($product['pack_items'] as $pack_item) {
+                $packItemType = !empty($pack_item['pack_items']) ? OrderProductForViewing::TYPE_PACK :
+                    OrderProductForViewing::TYPE_PRODUCT_WITHOUT_COMBINATIONS;
+                $packItemImagePath = isset($pack_item['image_tag']) ?
+                    $this->imageTagSourceParser->parse($pack_item['image_tag']) :
+                    null;
+                $packItems[] = new OrderProductForViewing(
```
Comment (PierreRambaud) [inline src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:453]: What a constructor :sweat_smile:

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php [hunk 3]
```diff
@@ -438,8 +438,42 @@ private function getOrderProducts(Order $order): OrderProductsForViewing
...
+                    $pack_item['id_product'],
+                    $pack_item['name'],
+                    $pack_item['reference'],
+                    $pack_item['supplier_reference'],
+                    $pack_item['pack_quantity'],
+                    0,
+                    0,
+                    $pack_item['current_stock'],
+                    $packItemImagePath,
+                    Tools::ps_round(0, 2),
```
Comment (atomiix) [inline src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php]: Shouldn't we be using `Context::getContext()->getComputingPrecision()` for the rounding precision here?

Comment (Progi1984) [reply src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php]: Done. Thanks

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php (hunk 2 above)
Comment (Progi1984) [reply src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:453]: :smile:

File: admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js (hunk 1 above)
Comment (Progi1984) [reply admin-dev/themes/new-theme/js/pages/order/OrderViewPageMap.js]: Done. Thanks

File: src/Core/Domain/Order/QueryResult/OrderProductForViewing.php [hunk 4]
```diff
@@ -195,6 +215,14 @@ public function getName(): string
         return $this->name;
     }
 
+    /**
+     * @return array<OrderProductForViewing>
```
Comment (atomiix) [inline src/Core/Domain/Order/QueryResult/OrderProductForViewing.php]: Same here?

File: src/Core/Domain/Order/QueryResult/OrderProductForViewing.php [hunk 5]
```diff
@@ -43,6 +49,11 @@ class OrderProductForViewing
      */
     private $name;
 
+    /**
+     * @var array<OrderProductForViewing>
```
Comment (atomiix) [inline src/Core/Domain/Order/QueryResult/OrderProductForViewing.php]: Is that a new convention we started to use? I'm used to see `OrderProductForViewing[]` instead 🤔

File: src/Core/Domain/Order/QueryResult/OrderProductForViewing.php (hunk 4 above)
Comment (Progi1984) [reply src/Core/Domain/Order/QueryResult/OrderProductForViewing.php]: Fixed

File: src/Core/Domain/Order/QueryResult/OrderProductForViewing.php (hunk 5 above)
Comment (Progi1984) [reply src/Core/Domain/Order/QueryResult/OrderProductForViewing.php]: Fixed

**Review by atomiix — DISMISSED** [review]
LGTM !

**Comment by Robin-Fischer-PS** [comment]
Hi @Progi1984 !

Four issues on this PR : 
- The link on each product is redirecting to the order instead of the product
- When a product has a quantity in the pack > 1 , I suppose it should have the same design as on the product block (https://user-images.githubusercontent.com/27484094/69815006-b637ee80-11f5-11ea-82d2-9936d0859367.png)
- There should be the text "Reference number :" before the Reference number
- Could the popin be larger ? (In order to make "Package item" fit in one line)

:)

**Comment by TristanLDD** [comment]
@Progi1984 I have a comment: the pack content popup is supposed to have the style of a back-office popup:

Capture d’écran 2020-01-21 à 18 16 59

Here I can see it has the style of a regular container with the behavior of a popup: 
image

**Comment by Progi1984** [comment]
@TristanLDD I use the default style of Boostrap Modal. May be the UI Kit is not uptodate on the BO ?

**Comment by TristanLDD** [comment]
> @TristanLDD I use the default style of Boostrap Modal. May be the UI Kit is not uptodate on the BO ?

I think it's because it's the first time we use modals with table inside. Would it be possible to customize it so it looks more like the mockups? If not, can we make it at least the same size as the mockups? 
Thanks

**Comment by Progi1984** [comment]
@TristanLDD In this cas, it must be done on the UI Kit side for not overriding that. (Poke @NeOMakinG)

**Comment by Progi1984** [comment]
@Robin-Fischer-PS 

- :ballot_box_with_check: The link on each product is redirecting to the order instead of the product
- :ballot_box_with_check: When a product has a quantity in the pack > 1 , I suppose it should have the same design as on the product block (https://user-images.githubusercontent.com/27484094/69815006-b637ee80-11f5-11ea-82d2-9936d0859367.png)
- :ballot_box_with_check: There should be the text "Reference number :" before the Reference number
- :ballot_box_with_check: Could the popin be larger ? (In order to make "Package item" fit in one line)

File: src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php (hunk 2 above)
Comment (matks) [reply src/Adapter/Order/QueryHandler/GetOrderForViewingHandler.php:453]: (nice-to-have) Could be improved by creating a dedicated constructor
```
$packItems[] = new OrderProductForViewing::createPackItem(...);
```

The constructor can either accepts all the arguments (at least we can remove the `null`) or a DTO

**Review by matks — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]
lgtm 👍

**Comment by Robin-Fischer-PS** [comment]
Hi ! It's OK for QA :heavy_check_mark: :)

Thanks @Progi1984 !

**Comment by lucaballerini** [comment]
What about Webservice? When retrieving order details and the order contains a Pack product we don't have details about the products in the pack chosen (neither which combinations) as we have instead in BO. Attached XML returned and Screen of BO
Schermata 2020-06-11 alle 15 25 41

[order67xml.txt](https://github.com/PrestaShop/PrestaShop/files/4765006/order67xml.txt)

**Pull request merged**